src/protocol/context/malicious.rs

```diff
@@ -24,36 +25,59 @@ use crate::secret_sharing::replicated::{
     malicious::AdditiveShare as MaliciousReplicated, semi_honest::AdditiveShare as Replicated,
 };
 use crate::secret_sharing::Arithmetic;
-use crate::sync::Arc;
 
 /// Represents protocol context in malicious setting, i.e. secure against one active adversary
 /// in 3 party MPC ring.
 #[derive(Clone, Debug)]
-pub struct MaliciousContext<'a, F: Field> {
-    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
-    /// may operate with raw references and be more efficient
-    inner: Arc<ContextInner<'a, F>>,
+pub struct MaliciousContext<'c, 'a, F: Field> {
+    inner: &'c ContextInner<'c, 'a, F>,
     step: Step,
     total_records: TotalRecords,
+    // Ensure that the struct is invariant over 'a, because that may be required in the future.
+    _dummy: PhantomData<OnceCell<&'a ()>>,
 }
 
-pub trait SpecialAccessToMaliciousContext<'a, F: Field> {
+/// Represents protocol context in malicious setting, i.e. secure against one active adversary
+/// in 3 party MPC ring.
+#[derive(Debug)]
+pub struct MaliciousContextBuf<'c, 'a, F: Field> {
+    inner: ContextInner<'c, 'a, F>,
+    step: Step,
+    total_records: TotalRecords,
```

Comment:
How much do we depend on these two fields (step/total_records) being available on this `...Buf` struct?  How often is this specific type used as opposed to the reference type?

It seems to me like this type could omit these fields and the `get_ref()` function could populate the values with a default/initial value.

Looking at how you use `get_ref()`, it might be the latter: you might pass a `Substep` implementation to the function.

Reply:
> How much do we depend on these two fields (step/total_records) being available on this ...Buf struct? How often is this specific type used as opposed to the reference type?

The answer to the second question is that the owned type is used rarely. I think I may have tried removing the step/records from the buf type at some point during development, but I can't find the version of the code where I had done that, nor can I remember why I decided not to. Possibly because SpecialAccessToMaliciousContext appeared to need to store step/records in the owned type? But I don't think that's a good reason, SpecialAccessTopMaliciousContext is a special case constructor and I've in fact already changed it to return a borrowed instead of owned context in the latest version of my changes. There were also periods where I thought that I might be able to have ContextBuf deref to Context or to otherwise have the test closures accept ContextBuf as their argument rather than Context, in which case ContextBuf might have more need to hold the full state of a context.

Another variant would be to eliminate the ContextBuf owned type entirely, and just use ContextStore as the owned type (or keep ContextBuf, and get rid of ContextStore).